### ISSUE TITLE
Promise detection

### ISSUE THREAD

**Issue by bplatz** [ldegen/dependableP#1]
Found the library, I like where it is going.

What do you think about looking for the presence of any Promise methods (like .then()) to 'auto-detect' (i.e. typeof token === 'object' && typeof token.then === 'function'. This is specifically in relation to you not having to create the new registerP and resolveP. I know it isn't a perfect detection mechanism but I think most people would like the 'magic'. You could also provide a way to disable the magic if someone didn't want it (see below).

Your forgeArgList 'hack' could also be used to indicate a promise by some convention, so long as you are going the annotation route anyhow.

I like the idea of a 'strict mode' option you can set on the library. By default you can auto-detect promises, auto-resolve dependencies by argument names and basically do a lot of magic. If the strict mode was enabled, no magic would be performed and things like promises would have to be explicitly declared via annotations. Perhaps even the injections would have to be annotated.

**Comment by ldegen** [comment]
Hm... I am not completely sure what problem you are trying to solve.
The `register`/`resolve` vs `registerP`/`resolveP` thing? Does it bother you that there are two APIs?

Sorry, I haven't been thinking about this project for some time now, so I just might a bit slow-minded. :-)

**Comment by bplatz** [comment]
Exactly, having separate methods to register. Having one method that does the right thing seems a lot nicer, with an option to be explicit if necessary.

**Comment by ldegen** [comment]
Yes, well, I somehow got stuck there. 

My original idea was to proceed morless like you suggested, i.e. to "override" the original API an provide transparent support for promises in addition to the original functionality. I thought I could do it in a way so that client code can be largely ignorant of the fact that promises are involved. I decided to use separate function names mostly to be on the save side until I was sure that everything worked as I intended (from the very beginning, I used this in customer's project, which I didn't want to break)

For the callback-based part of the original API, there is no problem. Actually, there isn't any need for guessing games here. The framework can lift and unlift simply _all_ arguments/return values of any callback, so if one factory decides to produce a promise, nobody else needs to even know about it. Very sweet and simple. I would really like to keep it this way.

Unfortunately, things like `get` break this monadic bliss. Code that uses `get` needs to be aware of promises. So strictly speaking, dependableP can never be an actual extension of dependable; it does not extend, but _change_ the original behaviour in an incompatible way.
I do not think this can be fixed by any cleverness in the `register`/`resolve`-Methods. The wrapper methods simply do a different thing than the original ones and as long as there is a `get` method, it is probably not possible to compensate for that. Overriding the original methods wouldn't help (much).

Since I ran out of time and since what I had was good enough for the projects I wanted to use it in, I had to let it go. So it wouldn't surprise me, if there was a simple solution that I haven't thought of.
